### ISSUE TITLE
Backup my Codeberg content

### ISSUE THREAD

**Issue by Benjamin-Loison** [Benjamin-Loison/gitea#46]
### Feature Description

As [my number of repositories increase request](https://matrix.to/#/!wkyjxWeAOBpKMvHbno:matrix.org/$RHZM-E8FsCAQPC8Sar9kDTYjvRUoiFcnaKEezO7TFgI) seems to have more impact than expected.

https://github.com/go-gitea/gitea/issues/1446

Can migrate whole user if have access to another Gitea instance?

Requires admin to proceed officially it seems:
- https://docs.gitea.com/next/administration/backup-and-restore
- https://docs.gitea.com/next/development/migrations-interfaces

Making mirrors only accepting cryptographically signed changes seems to be the perfect way of proceeding to not trust the possible hosting platform arbitrary actions.

Backup issues is the priority (hence consider search engine results for `gitea export issues`):
- https://github.com/go-gitea/gitea/issues/4808 only proposes a title export
- [`tea issues`](https://forum.gitea.com/t/hack-easiest-way-to-export-issues-to-csv/8038/2) none of `--output value, -o value      Output format. (csv, simple, table, tsv, yaml)` (source: `tea issues -h`) export issue details
- `tea issues --repo Firefox 40` except in actual terminal formatting, it does not seem to return Markdown one, so seems clear that cannot restore from that
- [python-gitea-backup](https://github.com/simplebackups/python-gitea-backup) [does not seem to work](https://gitea.lemnoslife.com/Benjamin_Loison/python-gitea-backup/issues/5)

Related to [YouTube-operational-API/issues/211](https://github.com/Benjamin-Loison/YouTube-operational-API/issues/211), #9 and [Benjamin_Loison/codeberg/issues/3](https://gitea.lemnoslife.com/Benjamin_Loison/codeberg/issues/3).

+619

**Comment by Benjamin-Loison** [comment]
Also note that this issue is important as Codeberg is sometimes facing access issue and is sometimes just slow, so this may be early signs of platform exhaustion.

**Comment by Benjamin-Loison** [comment]
Maybe similarly should have a backup of:
- https://invent.kde.org/benjaminloison
- https://code.videolan.org/Benjamin_Loison
- ...

![image](https://github.com/Benjamin-Loison/gitea/assets/12752145/1d0c32e6-69f3-4b44-8502-2b3dfb0f9222)

Source: https://invent.kde.org/benjaminloison/kile/-/issues/new

uBlock rule:

```
##.some.banner.red.gl-my-6.gl-broadcast-message
```

![image](https://github.com/user-attachments/assets/22a96b8e-f16f-4e48-a022-cdb7b3cee89b)

Source: https://gitlab.gnome.org

Related to [Improve_websites_thanks_to_open_source/issues/1275](https://codeberg.org/Benjamin_Loison/Improve_websites_thanks_to_open_source/issues/1275).

<details>
<summary>Also sometimes get <i>504 Gateway Time-out</i>:</summary>

```html
<html><body><h1>504 Gateway Time-out</h1>
The server didn't respond in time.
</body></html>
```
</details>

On 20/10/24 at 11:14. At 1:27 UTC+2 it was working fine as far as I remember according to my Linux Mint 22 Cinnamon Framework 13 Firefox history.

https://pad.gnome.org/wiki-migration-guide#gitlabgnomeorg so if it still fails after 25/10/24 then there is something to investigate.

No more issue as of 21/10/24 at 05:36.

<details>
<summary>As of 24/10/24 11:21 face again <i>504 Gateway Time-out</i> but also face <i>500: We're sorry, something went wrong on our end</i>:</summary>

```html
<!DOCTYPE html>
<html>
<head>
  <meta content="width=device-width, initial-scale=1" name="viewport">
  <title>Something went wrong (500)</title>
  <style>
    body {
      color: #333238;
      text-align: center;
      font-family: "Nunito Sans", -apple-system, ".SFNSText-Regular", "San Francisco", BlinkMacSystemFont, "Segoe UI", "Helvetica Neue", Helvetica, Arial, sans-serif;
    }

    h1 {
      font-size: 1.75rem;
      line-height: 2.25rem;
      margin: 1rem 0;
    }

    p {
      margin-bottom: .5rem;
    }

    img {
      display: block;
      margin: 0 auto;
    }

    a {
      text-decoration: none;
      color: #1068bf;
    }

    a:hover {
      text-decoration: underline;
    }

    .error-container {
      max-width: 65ch;
      margin: auto;
      padding: 1rem 0;
    }

    .action-container {
      margin-top: 1.5rem;
    }

    .go-back {
      display: none;
    }
  </style>
</head>

<body>
  <div class="error-container">
    <img src='[/-/error-illustrations/error-500-lg.svg](view-source:https://gitlab.gnome.org/-/error-illustrations/error-500-lg.svg)' alt="500 error"/>
    <h1>500: We're sorry, something went wrong on our end</h1>
    <p>Request ID: <code>01JAYX4GAABA90SBXDM9JD0R17</code></p>
    <p>Try refreshing the page, or going back and attempting the action again.</p>
    <p>Please contact your GitLab administrator if this problem persists.</p>
    <div class="action-container">
      <a href="javascript:history.back()" class="js-go-back go-back">Go back</a>
    </div>
  </div>
  <script>
    (function () {
      var goBack = document.querySelector('.js-go-back');

      if (history.length > 1) {
        goBack.style.display = 'inline';
      }
```
</details>

[freedesktop/wikis/home?version_id=d67d9877e5c648b4afbe0fbd25baa261194acf0c#how-can-i-contribute-to-an-existing-project-or-create-a-new-one](https://gitlab.freedesktop.org/freedesktop/freedesktop/-/wikis/home?version_id=d67d9877e5c648b4afbe0fbd25baa261194acf0c#how-can-i-contribute-to-an-existing-project-or-create-a-new-one)

Source: https://gitlab.freedesktop.org/users/sign_in

Related to [Benjamin_Loison/gitlab/issues/34#note_2214342628](https://gitlab.com/Benjamin_Loison/gitlab/-/issues/34#note_2214342628).

<details>
<summary>As of 26/11/24 at 15:02 from my 4G I got (but no more) for <a href="https://gitlab.gnome.org/Benjamin_Loison/gimp/-/issues/30">Benjamin_Loison/gimp/issues/30</a>:</summary>

```html
<?xml version="1.0" encoding="utf-8"?>
<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Strict//EN"
 "http://www.w3.org/TR/xhtml1/DTD/xhtml1-strict.dtd">
<html>
  <head>
    <title>503 No healthy backends</title>
  </head>
  <body>
    <h1>Error 503 No healthy backends</h1>
    <p>No healthy backends</p>
    <h3>Error 54113</h3>
    <p>Details: cache-cdg-XXXXXXXXXXX-CDG 1732629718 XXXXXXXX</p>
    <hr>
    <p>Varnish cache server</p>
  </body>
</html>
```
</details>

https://invent.kde.org on 04/12/24 at 11:14:

<details>
<summary><i>HTTP 502: Waiting for GitLab to boot</i>:</summary>

```html
<!DOCTYPE html>
<html>
<head>
  <meta content="width=device-width, initial-scale=1" name="viewport">
  <title>Waiting for GitLab to boot</title>
  <style>
    body {
      color: #333238;
      text-align: center;
      font-family: "Nunito Sans", -apple-system, ".SFNSText-Regular", "San Francisco", BlinkMacSystemFont, "Segoe UI", "Helvetica Neue", Helvetica, Arial, sans-serif;
    }

    h1 {
      font-size: 1.75rem;
      line-height: 2.25rem;
      margin: 1rem 0;
    }

    p {
      margin-bottom: .5rem;
    }

    img {
      display: block;
      margin: 0 auto;
    }

    a {
      text-decoration: none;
      color: #1068bf;
    }

    a:hover {
      text-decoration: underline;
    }

    .error-container {
      max-width: 65ch;
      margin: auto;
      padding: 1rem 0;
    }

    .action-container {
      margin-top: 1.5rem;
    }

    .go-back {
      display: none;
    }

    .footer {
      margin: 1rem 0 0 0;
      padding: 1rem 0 0 0;
      border-top: 1px solid #dcdcde;
      color: #737278;
      font-size: 0.75rem;
    }
    .footer p {
      margin: 0;
    }
  </style>
</head>

<body>
  <div class="error-container">
    <img src='[/-/error-illustrations/error-502-lg.svg](view-source:https://invent.kde.org/-/error-illustrations/error-502-lg.svg)' alt="502 error"/>
    <h1>HTTP 502: Waiting for GitLab to boot</h1>
    <p>It can take up to a few minutes for GitLab to boot completely.</p>
    <p>This page will automatically reload every 5 seconds.</p>
    <div class="action-container">
      <a href="javascript:history.back()" class="js-go-back go-back">Go back</a>
    </div>

    <div class="footer">
      <p>Refreshed at <span id="refreshedAt"></span></p>
    </div>
  </div>

  <script>
    const now = new Date();
    const refreshedAt = now.toLocaleTimeString();

    document.getElementById("refreshedAt").innerText = refreshedAt;

    window.setTimeout(() => location.reload(), 5000);

    if (history.length > 1) {
      const goBack = document.querySelector('.js-go-back');
      goBack.style.display = 'inline';
    }
  </script>
</body>
</html>
```
</details>

+5

**Comment by Benjamin-Loison** [comment]
Related to [Benjamin_Loison/git/issues/75](https://codeberg.org/Benjamin_Loison/git/issues/75).

**Comment by Benjamin-Loison** [comment]
Should investigate why the attachments of:
- [Benjamin_Loison/moby/issues/2#issue-770365](https://codeberg.org/Benjamin_Loison/moby/issues/2#issue-770365)
- [Benjamin_Loison/Debian/issues/10#issue-750788](https://codeberg.org/Benjamin_Loison/Debian/issues/10#issue-750788)
- [Robust_image_source_identification_on_modern_smartphones/issues/25#issuecomment-1495](https://gitea.lemnoslife.com/Benjamin_Loison/Robust_image_source_identification_on_modern_smartphones/issues/25#issuecomment-1495)

are gone.

**Comment by Benjamin-Loison** [comment]
I remember for sure that attachments of [Improve_websites_thanks_to_open_source/issues/234](https://codeberg.org/Benjamin_Loison/Improve_websites_thanks_to_open_source/issues/234) were working recently, maybe at *Jul 28, 2024, 9:37 PM GMT+2*, according to *edited* history.

While these attachments were debatable I was already having doubts on not debatable content, so let us consider this issue seriously concerning Codeberg where are my main contributions. Furthermore, I suspect that I would have been noticed if such content was not authorized and removed. Let us focus on issues, as only attachments of issues seem implied and it seems to indeed make sense they focus on that first. 

Let us consider as an example [Improve_websites_thanks_to_open_source/attachments/80d6de36-f4ae-45c6-a2b0-1e51ef8bf4a4](https://codeberg.org/Benjamin_Loison/Improve_websites_thanks_to_open_source/attachments/80d6de36-f4ae-45c6-a2b0-1e51ef8bf4a4):

```bash
curl 'https://codeberg.org/Benjamin_Loison/Improve_websites_thanks_to_open_source/attachments/80d6de36-f4ae-45c6-a2b0-1e51ef8bf4a4'
```
```
Not Found
```

```bash
curl -I 'https://codeberg.org/Benjamin_Loison/Improve_websites_thanks_to_open_source/attachments/80d6de36-f4ae-45c6-a2b0-1e51ef8bf4a4'
```
<details>
<summary>Output:</summary>

```
HTTP/2 404 
cache-control: max-age=0, private, must-revalidate, no-transform
content-type: text/plain; charset=utf-8
set-cookie: i_like_gitea=d4d711a41a026427; Path=/; HttpOnly; Secure; SameSite=Lax; Secure; SameSite=Lax
set-cookie: _csrf=XjgqAnsj9kH-RObBlcH7tCaOjdw6MTcyMjQ0MDcwMTc4MzU0NDEwNQ; Path=/; Max-Age=86400; HttpOnly; Secure; SameSite=Lax; Secure; SameSite=Lax
date: Wed, 31 Jul 2024 15:45:01 GMT
content-length: 10
strict-transport-security: max-age=63072000; includeSubDomains; preload
permissions-policy: interest-cohort=()
x-frame-options: sameorigin
x-content-type-options: nosniff
```
</details>

**Comment by Benjamin-Loison** [comment]
List given user repositories, notably those private.
List issues of each repository, all are public as far as I know.
Save for each issue all its comments Markdown and attachments. Could even just for the moment focus on attachments.

**Comment by Benjamin-Loison** [comment]
Can just save the attachment behind each attachment URL.

**Comment by Benjamin-Loison** [comment]
There does not seem to be an endpoint to directly list attachments of a user. What about issue comments or issues? It does not seem to be the case too according to `issue`, `user` and section search.

**Comment by Benjamin-Loison** [comment]
<details>
<summary>Python script making statistics and downloading all attachments still available:</summary>

```python
import requests
import json
from tqdm import tqdm
import os
import shutil
import re
import requests_cache

requests_cache.install_cache('requests_cache_codeberg')

# *Repository and Organization Access*: *Public only*
# *Selection permissions*:
# - *repository*: *Read*
# - *user*: *Read*
# - *issue*: *Read*
TOKEN = 'XXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX'
I_LIKE_GITEA_COOKIE = 'XXXXXXXXXXXXXXXX'
DOWNLOAD = True

HEADERS = {
    'Authorization': f'token {TOKEN}'
}

def getApi(url, params):
    return requests.get(f'https://codeberg.org/api/v1/{url}', params, headers = HEADERS).json()

def getItems(url, params):
    items = []
    with tqdm() as progressBar:
        while True:
            params['page'] = progressBar.n + 1
            data = getApi(url, params)
            items += data
            dataLen = len(data)
            if ('limit' in params and dataLen < params['limit']) or data == []:
                break
            progressBar.update(1)
    return items

params = {
    'limit': 50,
}

repos = getItems('user/repos', params)

ATTACHMENT_REGEX = re.compile('!\\[.*\\]\\(/attachments/([a-f0-9]{8}\\-[a-f0-9]{4}\\-[a-f0-9]{4}\\-[a-f0-9]{4}\\-[a-f0-9]{12})\\)')

uniqueAttachmentMentioned = 0
uniqueAttachmentMentionedButMissing = 0

def mkdir(path):
    os.makedirs(path, exist_ok = True)

def downloadAssets(assets, desc):
    if DOWNLOAD:
        for asset in tqdm(assets, desc = f'{desc} asset'):
            uuid = asset['uuid']
            filePath = f'attachments/{uuid}'
            if not os.path.isfile(filePath):
                with requests.get(asset['browser_download_url'], cookies = {'i_like_gitea': I_LIKE_GITEA_COOKIE}, stream = True) as r:
                    with open(filePath, 'wb') as f:
                        shutil.copyfileobj(r.raw, f)

def treatComment(comment, assets):
    global uniqueAttachmentMentioned, uniqueAttachmentMentionedButMissing
    assetsUuids = set(map(lambda asset: asset['uuid'], assets))
    commentUuids = set(ATTACHMENT_REGEX.findall(comment['body']))
    uniqueAttachmentMentioned += len(commentUuids)
    uniqueAttachmentMentionedButMissing += len(commentUuids - assetsUuids)

for repo in tqdm(repos, 'repo'):
    repoFullName = repo['full_name']
    issues = getItems(f'repos/{repoFullName}/issues', {'limit': 50, 'state': 'all'})
    for issue in tqdm(issues, 'issue'):
        assets = getApi(f'repos/{repoFullName}/issues/{issue["number"]}/assets', {})
        downloadAssets(assets, 'issue')
        treatComment(issue, assets)
    comments = getItems(f'repos/{repoFullName}/issues/comments', {})
    for comment in tqdm(comments, 'comment'):
        assets = comment['assets']
        downloadAssets(assets, 'comment')

print(f'{uniqueAttachmentMentioned=}')
print(f'{uniqueAttachmentMentionedButMissing=}')
print(f'{100 * (uniqueAttachmentMentionedButMissing / uniqueAttachmentMentioned)=}')
```
</details>

`treatComment` does not seem used for issue comments which does not seem correct.

**Comment by Benjamin-Loison** [comment]
<details>
<summary>Related to another issue:</summary>

```
-----BEGIN PGP MESSAGE-----

hF4DTQa9Wom5MBgSAQdAjJsEcbTXqHfROw1ogXMHR2OPdPW3d4HY6FDAYiRklGsw
FpV+4xpho4NGUL6oUZ+MFyJeq/Odo8OmzNWUHT+qvedrJs9JEp8JFHrdWS9Fzxws
0mwBh2+vjv+pK4SvQJBhLzLmaenes34vNzgvhwEyEQIYYZIXL52+cDGZPpfDlv9I
qkVAUkcYIbtHbrj5TbumnmpzcZJC2xSNX8oWS+BzzP+q9fmvQ4kJoQaj8Y82A//l
fZ4GbnlYX6wnsGxP2uE=
=9U8T
-----END PGP MESSAGE-----
```
</details>

**Comment by Benjamin-Loison** [comment]
Can either proceed with:
- `/repos/{owner}/{repo}/issues/{index}/assets`
- `/repos/{owner}/{repo}/issues/comments/{id}/assets`

These would result in different complexities.

Have to ensure that get attachments of initial comment and all the following ones.

[Example_to_backup_attachments_of_issues/issues/1](https://codeberg.org/Benjamin_Loison/Example_to_backup_attachments_of_issues/issues/1) shows that we need both:

<details>
<summary>Python script investigating API difference between first and following issue comments:</summary>

```python
REPOSITORY = 'Example_to_backup_attachments_of_issues'
INDEX = 1

print(json.dumps(getApi(f'repos/Benjamin_Loison/{REPOSITORY}/issues/{INDEX}'), indent = 4))

print(json.dumps(getApi(f'repos/Benjamin_Loison/{REPOSITORY}/issues/{INDEX}/assets'), indent = 4))

print(json.dumps(getApi(f'repos/Benjamin_Loison/{REPOSITORY}/issues/comments'), indent = 4))
```
</details>

**Comment by Benjamin-Loison** [comment]
For removed attachments, could search with a regex the different UUIDs and check if we have them attached.

Should also backup Codeberg attachments that have been uploaded previously as may have censored them, see #12.

**Comment by Benjamin-Loison** [comment]
I retrieved 1,295 files for 787.8 MB according to `nemo`.

See `/home/benjamin/Desktop/bens_folder/dev/git/codeberg/attachments/` on my Linux Mint 22 Cinnamon Framework 13.

**Comment by Benjamin-Loison** [comment]
<details>
<summary>Issue comment source with missing attachments:</summary>

```python
S = '''https://www.youtube.com/watch?v=vEIpNuCh35Q

Could propose tool for channels to list them automatically.
Based on avatar, as they do not even change them.
What about the messages?
Especially when the video is published as some bots are already working at this moment and there is no other comments.

https://www.youtube.com/watch?v=bpeI2PwnFVc
![image](/attachments/179afbf0-ca7b-4f67-941e-cd03132eb4b9)

https://www.youtube.com/watch?v=bpeI2PwnFVc&lc=UgzlaZOeSWhXi5tBCDR4AaABAg
https://www.youtube.com/channel/UCqKfyQbuWL_u7Zvl9ZOe5gQ
https://yt3.ggpht.com/CLtfnJ9CmqVnVvY38x0mNKCSGC5T4pjsbIUG2TFOXYLw7zSOHKcUq98dLHgLSm1Juq5yqE-NAtE=s88-c-k-c0x00ffffff-no-rj
![image](/attachments/1b2a8a1a-40a8-4f05-8cce-b1392f1f3292)
`@IsabelleFulleri`
`Vos vidéos sont comme un livre avec un contenu clair. Très pratique pour les spectateurs et facile à comprendre. ✋ & 🐽`
![image](/attachments/99a02d72-d255-47f7-96b5-692dfbfae846)
![image](/attachments/80d6de36-f4ae-45c6-a2b0-1e51ef8bf4a4)
![image](/attachments/d4ab7344-4288-44e2-bf1a-f61ceca65643)

https://www.youtube.com/watch?v=bpeI2PwnFVc&lc=UgzJdf0bMGm_V1T50jp4AaABAg
https://www.youtube.com/channel/UCpE32kLG6Ygb34lO3H6O9MA
https://yt3.ggpht.com/XEM_T0wFR4htx1iIZA-m0eRlzONRiHvrqIWodaDiSzB9HK8nFIUV7arBqGX9kg5POYTYa10A=s88-c-k-c0x00ffffff-no-rj
![image](/attachments/ce508c89-9de2-4f7d-9e36-b56583c578b4)
`@NathalieGarcia19`
`Vous créez du matériel intéressant pour les débutants et les spectateurs expérimentés. Bonne polyvalence. 😁😁 8 🛹`

After some time:

![image](/attachments/6b4be8d6-58ae-428e-a557-80b62c4c9109)

Let us wait another nudify spam comment on another video to see if above ones are deleted.

Note that after 22 hours the nudity comments are still there.

Someone on [Discord](https://discord.com/channels/933841502155706418/933841503103627316/1186712630862291008) is interested by this exact issue.

> Personally I recognize the spammer account thumbnails from a video to the other, hence I think we can identify them this way (assuming nobody use the spammer thumbnails).

> Also something important I noticed is that some spams are posted right away after the video publication, it can help detecting them.

> I recommend [Subscribe to Push Notifications](https://developers.google.com/youtube/v3/guides/push_notifications) to be notified of video uploads.

> In my opinion we should provide an open-source tool such that YouTube creators can opt-in and automatically get rid of such comments, it does not seem that difficult to do.

[I asked him news](https://discord.com/channels/933841502155706418/933841503103627316/1195430589025755187), let us wait a week, i.e. January 19 2024, before continuing on my own.

See [his June 17, 2024 update](https://discord.com/channels/933841502155706418/933841503103627316/1252159092139626547).

[Best complexity approach](https://discord.com/channels/933841502155706418/933841503103627316/1186716355844644976).

+8'''
```
</details>

Source: [Improve_websites_thanks_to_open_source/issues/234](https://codeberg.org/Benjamin_Loison/Improve_websites_thanks_to_open_source/issues/234)

**Comment by Benjamin-Loison** [comment]
```
uniqueAttachmentMentioned=1027
uniqueAttachmentMentionedButMissing=232
100 * (uniqueAttachmentMentionedButMissing / uniqueAttachmentMentioned)=22.59006815968841
```

Could be interesting to execute again to ensure these results but currently I get:

> :(
> You sent too many requests to Codeberg.
> 
> We are sorry for the inconvenience, but the wild west situation around the AI gold rush has prompted many big players to "simply crawl the web". Their massive resources can cause traffic that is much more than Codeberg can handle. It was a difficult decision for us, but we started to block suspicious traffic more aggressively to keep Codeberg available and performant to most of our users. And your traffic was also suspicious to us.
> 
> If you run into this error for the first time, wait about half an hour and try again. If you keep running into this with your normal usage, please contact us so that we can improve our rate limiting. Write to us: [email codeberg@codeberg.org](mailto:codeberg@codeberg.org), [Mastodon](https://social.anoxinon.de/@codeberg) or our [Matrix channel](https://matrix.to/#/#codeberg.org:matrix.org).
> 
> Thank you for your understanding.

Note that some attachments may not be mentioned in the Markdown `body`.

**Comment by Benjamin-Loison** [comment]
So they delete some of my stuff without noticing me??

What are Codeberg storage limits in theory?

**Comment by Benjamin-Loison** [comment]
On my Debian 12 GNOME *Pegasus* I store the attachments I add now on at `/home/benjamin_loison/Desktop/BensFolder/DEV/codeberg/`.

On my Linux Mint 22 Cinnamon Framework 13 I use `/home/benjamin/Desktop/bens_folder/dev/git/codeberg/attachments/`.

Contrarily to *Save Link As...*, `wget` saves the attachment with its UUID. Otherwise can use *Copy Link* but it is not user-friendly with many attachments.

Note that I save intermediary attachments while my study focus on current attachments, hence have to pay attention in counting attachments.

Can also leverage Firefox shortcuts once right-clicked <kbd>o</kbd> for *Copy Image Link* and <kbd>v</kbd> for *Save Image As...*.

**Comment by Benjamin-Loison** [comment]
I doubt edit history to be exhaustive, for instance [Improve_websites_thanks_to_open_source/issues/153#issue-621074](https://codeberg.org/Benjamin_Loison/Improve_websites_thanks_to_open_source/issues/153#issue-621074).

**Comment by Benjamin-Loison** [comment]
<details>
<summary>Python script to download issue comment attachments:</summary>

```python
# [Benjamin-Loison/gitea/issues/46](https://github.com/Benjamin-Loison/gitea/issues/46)

import requests
from tqdm import tqdm
import os
import shutil

REPOSITORY_FULL_NAME = 'Benjamin_Loison/android'
ISSUE_ID = None
COMMENT_ID = 2165179
ATTACHMENTS_FOLDER = '/home/benjamin/Desktop/bens_folder/dev/git/codeberg/attachments/'

def getApi(url):
    return requests.get(f'https://codeberg.org/api/v1/{url}').json()

def downloadAssets(assets):
    for asset in tqdm(assets, desc = 'asset'):
        uuid = asset['uuid']
        filePath = os.path.join(ATTACHMENTS_FOLDER, uuid)
        if not os.path.isfile(filePath):
            with requests.get(asset['browser_download_url'], stream = True) as r:
                with open(filePath, 'wb') as f:
                    shutil.copyfileobj(r.raw, f)
        else:
            print(f'{uuid} already exists!')

if ISSUE_ID is not None:
    assets = getApi(f'repos/{REPOSITORY_FULL_NAME}/issues/{ISSUE_ID}/assets')
if COMMENT_ID is not None:
    comment = getApi(f'repos/{REPOSITORY_FULL_NAME}/issues/comments/{COMMENT_ID}')
    assets = comment['assets']

downloadAssets(assets)
```
</details>

See Linux Mint 22 Cinnamon Framework 13 `/home/benjamin/Desktop/bens_folder/dev/python/projects/backup_gitea_issue_attachments/main.py` to ease this script usage.

Related to [Improve_websites_thanks_to_open_source/issues/569](https://codeberg.org/Benjamin_Loison/Improve_websites_thanks_to_open_source/issues/569).

Would be nice to have an algorithm to treat all comments of an issue, as I have doubts concerning [Improve_websites_thanks_to_open_source/issues/967](https://codeberg.org/Benjamin_Loison/Improve_websites_thanks_to_open_source/issues/967). Furthermore, it may make this algorithm easier to run as providing an issue id is simpler than a comment id.

`os.path.join` is to support both Linux `/` and Windows `\`, source: [the Stack Overflow answer 2953843](https://stackoverflow.com/a/2953843).

**Comment by Benjamin-Loison** [comment]
Alternatively could precise the repository full name and issue id and find which attachments are missing but that would not help that much.

**Comment by Benjamin-Loison** [comment]
<details>
<summary><i>You've been rate limited screenshot</i>:</summary>

![image](https://github.com/user-attachments/assets/fb280cd8-4e4f-45c2-a7fa-fcb54a6659d2)
</details>

Trying again does not help.

While I am still able to load [Improve_websites_thanks_to_open_source/issues/new](https://codeberg.org/Benjamin_Loison/Improve_websites_thanks_to_open_source/issues/new).

When HTTPS method *POST* I get:

```json
{"errorMessage":"You've been rate limited","renderFormat":"text"}
```

Note that the issue is not created.

No DuckDuckGo and Google results for `"You've been rate limited" "codeberg.org"`.

Editing comments of issues still work fine while publishing a new one is not possible.

What about the API? Would be quite fine as only needed for new issues, not editing them. So could create quite empty ones. Even if it makes their histories less clear.

Do my total number of issues on the repository (968) and for my user (3,397, source: https://codeberg.org/issues) look like a threshold? No.

How much do I have to wait to not suffer anymore of this issue. Maybe it is because I was creating 12 issues quite fast at [Overclock3000_notifications/issues](https://codeberg.org/Benjamin_Loison/Overclock3000_notifications/issues) as I was turning my notes into issues.

Do not face anymore this issue after about 15 minutes.

I face again this issue after tens of minutes.

https://codeberg.org/api/swagger#/issue/issueCreateIssue

https://codeberg.org/user/settings/applications
*Token name* `Create issues (https://github.com/Benjamin-Loison/gitea/issues/46)`
*Repository and Organization Access* *Public only*
*issue* *Read and write*

<details>
<summary>Curl request:</summary>

```bash
 curl -X 'POST' \
  'https://codeberg.org/api/v1/repos/Benjamin_Loison/OverClock3000_notifications/issues' \
  -H 'Authorization: token XXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX' \
  -H 'accept: application/json' \
  -H 'Content-Type: application/json' \
  -d '{
  "title": "Test"
}'
```
</details>
<details>
<summary>Output:</summary>

```json
{
    "id": 886526,
    "url": "https://codeberg.org/api/v1/repos/Benjamin_Loison/OverClock3000_notifications/issues/19",
    "html_url": "https://codeberg.org/Benjamin_Loison/OverClock3000_notifications/issues/19",
    "number": 19,
    "user": {
        "id": 74932,
        "login": "Benjamin_Loison",
        "login_name": "",
        "source_id": 0,
        "full_name": "Benjamin Loison",
        "email": "benjamin.loison@orange.fr",
        "avatar_url": "https://codeberg.org/avatars/f2a71b4f7a73663f685cb8337d73698f",
        "html_url": "https://codeberg.org/Benjamin_Loison",
        "language": "en-US",
        "is_admin": false,
        "last_login": "2024-09-20T14:53:43Z",
        "created": "2022-12-05T15:13:33Z",
        "restricted": false,
        "active": true,
        "prohibit_login": false,
        "location": "",
        "pronouns": "",
        "website": "",
        "description": "",
        "visibility": "public",
        "followers_count": 1,
        "following_count": 1,
        "starred_repos_count": 3,
        "username": "Benjamin_Loison"
    },
    "original_author": "",
    "original_author_id": 0,
    "title": "Test",
    "body": "",
    "ref": "",
    "assets": [],
    "labels": [],
    "milestone": null,
    "assignee": null,
    "assignees": null,
    "state": "open",
    "is_locked": false,
    "comments": 0,
    "created_at": "2024-09-26T11:40:26Z",
    "updated_at": "2024-09-26T11:40:26Z",
    "closed_at": null,
    "due_date": null,
    "pull_request": null,
    "repository": {
        "id": 262035,
        "name": "OverClock3000_notifications",
        "owner": "Benjamin_Loison",
        "full_name": "Benjamin_Loison/OverClock3000_notifications"
    },
    "pin_order": 0
}
```
</details>

[OverClock3000_notifications/issues/19](https://codeberg.org/Benjamin_Loison/OverClock3000_notifications/issues/19) works but does it work during the web interface issue? I deleted the created issue as it was only for testing purpose.

<details>
<summary>Related to issue:</summary>

```
-----BEGIN PGP MESSAGE-----

hF4DTQa9Wom5MBgSAQdAI2PAW6OcuBJrHOJgJ9J8AWMEmbmf5QiJkLL+VpA/UHkw
5t/2mQhrUbArZNCVY76ttcGonNvDjklWgzeTgVSRVyxGTCLy0QogkAbhNaWHPCfx
0mwBqI+fh1wuETq1ygEXSUmfXGhkBqIIynNhB5k5NxC7770ByYilINqDsnBmgfUq
O3hRSTfC2iXd6QRpF1anICRlcDa4wD8Cv6iTDcIc9UL0lKzbtP2UvY+PxeOU4Dmk
gRa88qDBk2/MaD3LsOE=
=4dtC
-----END PGP MESSAGE-----
```
</details>

I confirm that the web interface issue occurs before and after my working API request. At least without refreshing the webpage. Should test refreshing the webpage before and after. Note that if we refresh we lose to some extent the draft, so *Duplicate Tab* instead.

<details>
<summary>Python script equivalent to curl request:</summary>

```python
# [Benjamin-Loison/gitea/issues/46#issuecomment-2376547127](https://github.com/Benjamin-Loison/gitea/issues/46#issuecomment-2376547127)
import requests

TOKEN = 'XXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX'
REPOSITORY = 'YouTube_captions_search_engine'
TITLE = 'Test'

HEADERS = {
    'Authorization': f'token {TOKEN}',
    # Are both necessary?
    'accept': 'application/json',
    'Content-Type': 'application/json',
}

data = {
    'title': TITLE,
}

response = requests.post(
    f'https://codeberg.org/api/v1/repos/Benjamin_Loison/{REPOSITORY}/issues',
    headers = HEADERS,
    json = data,
)
```
</details>

I store it with `TOKEN` set on my Linux Mint 22 Cinnamon Framework 13 at `/home/benjamin/Desktop/bens_folder/dev/python/projects/gitea_create_issue/main.py` to use it faster.

Last API request took 134 seconds according to `print(response.elapsed.total_seconds())`, source: [the Stack Overflow answer 43260678](https://stackoverflow.com/a/43260678).

**Comment by Benjamin-Loison** [comment]
#9 API requests are taking unusual amount of time for [Improve_websites_thanks_to_open_source](https://codeberg.org/Benjamin_Loison/Improve_websites_thanks_to_open_source):

```
Closed issues page: 1it [04:30, 270.28s/it]
Open issues page:   5%|▌         | 1/19 [02:15<40:38, 135.45s/it]
```

**Comment by Benjamin-Loison** [comment]
It seems that as expected when Codeberg is down, its API is down as well.

**Comment by Benjamin-Loison** [comment]
<details>
<summary>Python script to publish issues due to rate limit:</summary>

```python
# [Benjamin-Loison/gitea/issues/46#issuecomment-2376547127](https://github.com/Benjamin-Loison/gitea/issues/46#issuecomment-2376547127)
import requests
import json
import pyperclip
import os

TOKEN = 'XXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX'
REPOSITORY = 'Improve_websites_thanks_to_open_source'
TITLE = '[discord.com] What are the formats of the various URLs that I use?'
USE_CLIPBOARD_AS_BODY = True

HEADERS = {
    'Authorization': f'token {TOKEN}',
    # Is the following necessary?
    'Content-Type': 'application/json',
}

data = {
    'title': TITLE,
}

# Using a variable prevent the cnsidered clipboard content being modified, hence potentially leaked between verifying clipboard content and publishing it.
clipboardContent = pyperclip.paste()
print(clipboardContent)

if USE_CLIPBOARD_AS_BODY:
    data['body'] = clipboardContent

def createIssue(repository, data):
    response = requests.post(
        f'https://codeberg.org/api/v1/repos/Benjamin_Loison/{repository}/issues',
        headers = HEADERS,
        json = data,
    )
    return response

response = createIssue(REPOSITORY, data)
print(response.text)
#print(json.dumps(response.json(), indent = 4))
#print(json.dumps(data, indent = 4))
TO_PUBLISH_FOLDER = 'to_publish'
data['repository'] = REPOSITORY
with open(f'{TO_PUBLISH_FOLDER}/{len(os.listdir(TO_PUBLISH_FOLDER))}.json', 'w') as f:
    json.dump(data, f, indent = 4)

for issueFileName in os.listdir(TO_PUBLISH_FOLDER):
    filePath = f'{TO_PUBLISH_FOLDER}/{issueFileName}'
    with open(filePath) as f:
        data = json.load(f)
    repository = data['repository']
    del data['repository']
    createIssue(repository, data)
    # Should verify publication before deleting or at least move to another completed folder.
    #os.remove(filePath)
```
</details>
<details>
<summary>Output:</summary>

```json
{"message":"NewIssue: \"Benjamin_Loison\" posted 32 issues in under 24 hours: rate limited","url":"https://codeberg.org/api/swagger"}
{"message":"","url":"https://codeberg.org/api/swagger"}
```
</details>

So it seems that the API is no more a workaround.

<details>
<summary>Note that <code>response.json()</code> returns:</summary>

```
Traceback (most recent call last):
  File "/home/benjamin/Desktop/bens_folder/dev/python/projects/gitea_create_issue/main.py", line 34, in <module>
    print(json.dumps(response.json(), indent = 4))
  File "/home/benjamin/venv/lib/python3.12/site-packages/requests/models.py", line 978, in json
    raise RequestsJSONDecodeError(e.msg, e.doc, e.pos)
requests.exceptions.JSONDecodeError: Extra data: line 2 column 1 (char 134)
```
</details>

So pending creation of issues are on my Linux Mint 22 Cinnamon Framework 13 at `/home/benjamin/Desktop/bens_folder/dev/python/projects/gitea_create_issue/to_publish/`.

Is it Codeberg specific or not? To see if user exception (unlimited amount of issues per period of time or higher one) has already been implemented.

```bash
jq -r '.body' 0.json
```
<details>
<summary>Output:</summary>

```
To simplify them at rendering.

*User ID*: `163399075771842560` is self standing
*Channel ID*: `1089261136374927511` is self standing as well
Well ids are all self standing I guess, however the URLs might not.
*Guild ID*: `933841502155706418`
*Message ID*: `1294417310337077259`

YouTube operational API `#annoucements` channel:
https://discord.com/channels/933841502155706418/1089261136374927511
so:
https://discord.com/channels/GUILD_ID/CHANNEL_ID

YouTube operational API `#annoucements` channel message:
https://discord.com/channels/933841502155706418/1089261136374927511/1294417310337077259
so
https://discord.com/channels/GUILD_ID/CHANNEL_ID/MESSAGE_ID

Message (`1294390509715980318`) with *Test de Benjamin Loison* (`811942490185793537`):

https://discord.com/channels/@me/811943803685437451/1294390509715980318

So no *User ID* of any participant of this 1 to 1 discussion. Would be unclear which one to work for both participants to refer to the same message, maybe the first person to start the discussion.

Another message URL within this discussion is:

https://discord.com/channels/@me/811943803685437451/1283794953184612363

so it seems to be

https://discord.com/channels/@me/CHANNEL_ID/MESSAGE_ID
```
</details>

**Comment by Benjamin-Loison** [comment]
Automatize publishing them once.

Should list the times when I published last issues then, also considering private repositories I guess. To plan when I can publish the issues automatically without proceeding regularly.

**Comment by Benjamin-Loison** [comment]
To put reference before issue creation, being able to ensure what issue id the new issue would have, possibly by using id of incorrectly created issue. However, I doubt that we can choose arbitrarily issue id at least concerning smaller ids. So could just delete the history but I suspect meta motification being shown between issue comments to stay.

**Comment by Benjamin-Loison** [comment]
Could just ask them what is their approach concerning data portability of GDPR.

**Comment by Benjamin-Loison** [comment]
DuckDuckGo and Google *Codeberg data portability* do not bring interesting results except below one.

[forgejo/issues/248](https://codeberg.org/forgejo/forgejo/issues/248) redirects to https://codeberg.org/forgejo/discussions/issues and no relevant issue is returned for *portability*.

**Comment by Benjamin-Loison** [comment]
Quite blocked by [Benjamin_Loison/requests-cache/issues/1](https://codeberg.org/Benjamin_Loison/requests-cache/issues/1).

**Comment by Benjamin-Loison** [comment]
We may think that relying on `requests_cache` may not be optimized for storing issues but in theory except the ridiculous redundant request information (if not even detected and compressed), it is quite optimized.

**Comment by Benjamin-Loison** [comment]
Quite blocked by [Benjamin_Loison/requests-cache/issues/2](https://codeberg.org/Benjamin_Loison/requests-cache/issues/2).

**Comment by Benjamin-Loison** [comment]
As Codeberg is down, should use own Gitea instance and push in realtime or even pull if have feedback but I doubt, or not, have such in the near future.
Including issues.
I doubt that I can implement safely inter instance mutex or maybe can leverage a private repository on Codeberg.
If able to have realtime notification of feedback then have an issue only if I write during the pulling latency which is for the moment not probable in my opinion, if it happens, then halt the synchronization and notify me for manual resolution.

**Comment by Benjamin-Loison** [comment]
Note that now have synchronized, even if Codeberg has no downtime, should keep updating my backup efficiently, preferably in realtime.

**Comment by Benjamin-Loison** [comment]
Related to [go-gitea/gitea/issues/18369](https://github.com/go-gitea/gitea/issues/18369).

My backup can consist in a dedicated instance/gitea user on another instance.

**Comment by Benjamin-Loison** [comment]
Tracked at [Benjamin_Loison/Food/issues/48](https://codeberg.org/Benjamin_Loison/Food/issues/48).

**Comment by Benjamin-Loison** [comment]
https://blog.codeberg.org/more-power-for-you-what-a-storage-quota-will-bring.html

> To make abuse prevention easier, we are applying some default storage limits on Codeberg. Free/libre software projects that need more will have to [kindly request more resources](https://codeberg.org/Codeberg-e.V./requests).

> Next to a lot of obvious storage abuse like encrypted personal backups

> All data on Codeberg probably fits into a 12 TB drive.

> Storage needs to be secure: While you might have backups, we don't want you to need them. Data is replicated, and backed up offsite, which requires more disks, and network bandwidth.

> The recent release of Forgejo v9 comes with a very important feature: Storage quotas.

> We are currently testing the feature and will enable it soon, carefully tweaking it to the needs of our users.

> Over the course of several months, we will slowly tune the quota rules. Currently, the following limits will apply to Codeberg users and orgs by default:

> - Git repository storage: 750 MiB
> - LFS, Packages, Releases, Attachments: Additional 1.5 GiB
> - Additionally, the total of "non-promoted" content, that is private and personal repositories ("my website", "my dotfiles"), should not exceed 100 MiB.

> If you know that you require more storage than we grant by default, kindly let us know via the [request procedures](https://codeberg.org/Codeberg-e.V./requests).

> If you exceed the limits and have not requested an exception, you will not be able to upload to Codeberg. There are currently no plans for removing existing content.

> We will be careful in adjusting the limits, aiming for zero friction for those who use our service.

> Do not hesitate to contact us with any questions or ideas you might have.

> In the past week, Codeberg has faced significant slowness and some downtime, most of which originates from illegitimate traffic such as new piracy content, excessive crawling and other activity that does not come from members of the free/libre software movement.

https://forgejo.org/2024-10-release-v9-0/

> Forgejo has chosen to use a “soft” quota implementation. It means that Forgejo checks the quota usage only before an action is executed, but it will allow a started action to complete.

> In some cases (like pushing to Git repositories), it is hard to estimate the exact new size, because it depends on how much data is available and how much we can benefit from compression. As a result, it is possible to exceed the quota if the operation was started before the quota was used up. After the quota is exceeded, new operations that would increase the quota won’t be possible.

> Furthermore, there is currently little support for early prevention of operations in the UI: The handling of, for example, web operations that are denied later is not yet optimal.

> https://forgejo.org/docs/v9.0/admin/quota/

**Comment by Benjamin-Loison** [comment]
Related to [Benjamin_Loison/gpg/issues/7](https://codeberg.org/Benjamin_Loison/gpg/issues/7).

**Comment by Benjamin-Loison** [comment]
Faced *You've been rate limited* pop up, as [#issuecomment-2376547127](#issuecomment-2376547127), again but workaround by not providing a body, hence just the issue title.

**Comment by Benjamin-Loison** [comment]
Note that I have saved some attachments that I have not published, so possibly containing secrets.

**Comment by Benjamin-Loison** [comment]
I may have saved attachments of https://gitea.lemnoslife.com by mistake.

**Comment by Benjamin-Loison** [comment]
> the total of "non-promoted" content, that is private [...] repositories [...], should not exceed 100 MiB

[PRNU_extraction](https://codeberg.org/Benjamin_Loison/PRNU_extraction) states 1.2 MB but as far as I know it does not count issues.

> git: 806 KiB, lfs: 415 KiB

[Benjamin_Loison/PRNU_extraction/settings](https://codeberg.org/Benjamin_Loison/PRNU_extraction/settings) also states 1.2 MB
[Benjamin_Loison/PRNU_extraction/settings/lfs](https://codeberg.org/Benjamin_Loison/PRNU_extraction/settings/lfs) shows 2 files.
[Benjamin_Loison/PRNU_extraction/settings/units#issues](https://codeberg.org/Benjamin_Loison/PRNU_extraction/settings/units#issues) does not give details.
What matters are attachments in issues.
Note that we cannot *Sort* issues by size.

**Comment by Benjamin-Loison** [comment]
<details>
<summary>Python script computing total size of issue attachments of a repository:</summary>

```python
#!/usr/bin/python3

# [Benjamin-Loison/gitea/issues/9](https://github.com/Benjamin-Loison/gitea/issues/9)

import requests
import json
from tqdm import tqdm
import math

GIT_INSTANCE_URL = 'https://codeberg.org'
# With `read:{issue,repository}` scopes.
TOKEN = 'XXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX'
REPOSITORY_FULL_NAME = 'Benjamin_Loison/Improve_websites_thanks_to_open_source'

headers = {
    'Authorization': f'token {TOKEN}'
}

def getApi(url, params = {}):
    response = requests.get(f'{GIT_INSTANCE_URL}/api/v1/{url}', params, headers = headers)
    return response.json()

LIMIT = 50
repositoryIssueParams = {
    'limit': LIMIT,
}
totalSizeRepositoryIssueAttachments = 0

repository = getApi(f'repos/{REPOSITORY_FULL_NAME}')
if repository.get('message') == 'The target couldn\'t be found.':
    print(f'Repository {GIT_INSTANCE_URL}/{REPOSITORY_FULL_NAME} not found!')
    exit(1)

def getSizeOfAssets(assets):
    return sum([asset['size'] for asset in assets])

# See [Benjamin-Loison/gitea/issues/60](https://github.com/Benjamin-Loison/gitea/issues/60).
openIssuesCount = repository['open_issues_count']
for state, total in [('closed', None), ('open', math.ceil(openIssuesCount / LIMIT))]:
    repositoryIssueParams['state'] = state
    with tqdm(total = total, desc = f'{state.title()} issues page') as progressBar:
        while True:
            repositoryIssueParams['page'] = progressBar.n + 1
            repositoryIssuePage = getApi(f'repos/{REPOSITORY_FULL_NAME}/issues', repositoryIssueParams)
            if repositoryIssuePage == []:
                break
            for repositoryIssue in repositoryIssuePage:
                assets = repositoryIssue['assets']
                totalSizeRepositoryIssueAttachments += getSizeOfAssets(assets)
            progressBar.update(1)

issueComments = getApi(f'repos/{REPOSITORY_FULL_NAME}/issues/comments')
for issueComment in issueComments:
    assets = issueComment['assets']
    totalSizeRepositoryIssueAttachments += getSizeOfAssets(assets)

print(f'Total size of repository issue attachments: {totalSizeRepositoryIssueAttachments:,} bytes')
```
</details>

```
Total size of repository https://codeberg.org/Benjamin_Loison/Improve_websites_thanks_to_open_source issue attachments: 325,261,219 bytes
```

Well 1 GB is too limited for private Git usage for my thesis as I massively use issues notably with attachments, so let us use https://gitea.lemnoslife.com.

Anyway should compute how much space I use with my private repositories.

[Benjamin-Loison/gitea/issues/145](https://github.com/Benjamin-Loison/gitea/issues/145) would help.

**Comment by Benjamin-Loison** [comment]
https://gitea.lemnoslife.com/admin/repos?sort=reversesize

**Comment by Benjamin-Loison** [comment]
https://codeberg.org/issues states 3564 (open) + 312 (closed) = 3876 issues.
https://codeberg.org/api/v1/repos/Benjamin_Loison/Improve_websites_thanks_to_open_source/issues?limit=1 seems to state that [Improve_websites_thanks_to_open_source/issues/1067](https://codeberg.org/Benjamin_Loison/Improve_websites_thanks_to_open_source/issues/1067) is 914900th issue n Codeberg so I have open 0.42 % of Codeberg issues.

https://codeberg.org/api/v1/repos/Benjamin_Loison/Improve_websites_thanks_to_open_source/issues/comments?limit=1
Can't easily know how many comments of issues I have published.

**Comment by Benjamin-Loison** [comment]
Migrate with API repositories to Codeberg.

**Comment by Benjamin-Loison** [comment]
Related to [Benjamin_Loison/linux/issues/20](https://gitea.lemnoslife.com/Benjamin_Loison/linux/issues/20).

**Comment by Benjamin-Loison** [comment]
<details>
<summary>Should also backup the Gitea instance:</summary>

```
-----BEGIN PGP MESSAGE-----

hF4DTQa9Wom5MBgSAQdApp1cEgxK79D9cQYHpwNYBqr24UNx3DDB5k7lFWVJb34w
HBVjcSdF4I3BD3pvm6sK72SvxwI6Sqbbp+JGfNyNgF7vl+EsJnUAAiid1yPVEFy0
0j8BpvLoZthL2IonrNxwqjBUXv4B8fr214ISlDK7wlBjio/y5cqcMdVDekiIcV6v
28P/eemiPxL8tdRSTh6bu60=
=Q/qV
-----END PGP MESSAGE-----
```
</details>

<details>
<summary>See the message:</summary>

```
-----BEGIN PGP MESSAGE-----

hF4DTQa9Wom5MBgSAQdAnwCUAzFr13HFYuw/eNi57IMZCBgjwHlZ+yuiujPGJ1Uw
Jl9wV6TZLmZOZkriUW9L2BlAufYn15pNWvLNSOwpw2Wf4oEDjcDQ3vthtnqAaIV8
0qoBVp5kF/P0fPjr5NZGzzxc950NqpfIdln4w4NXdhNUoTUT9YBTITRcyMD//0w5
x6J2mZVzRjJ4dJHITGSz93dCNuN6GRqAh6xyXAr0kaR9FUur0ITPgEIkqQU4qW5t
KNhw7uJ1ruc902aVmSwzRQPyC4U+66TOLW17xWPb+Hg4JgMpUnkOWcU2raPGUEQR
bxIrbCOajPHBRO/5h+1fv66b1K+t4UtxM07uUQ==
=1YS4
-----END PGP MESSAGE-----
```
</details>

**Comment by Benjamin-Loison** [comment]
Note that a solution should support somehow `git push --force`.

**Comment by Benjamin-Loison** [comment]
<details>
<summary>The person:</summary>

```
-----BEGIN PGP MESSAGE-----

hF4DTQa9Wom5MBgSAQdA0VvEEUt0Dw1Z/6JWRhyzn3MtFM59kZn6MnTOdzqhJnAw
HghxyGNkZZ7Qu3gZ97YvIykBsGE/sshuLqdj88i1hKvJxvUtNAajpiDqRt3upDFo
0kIB/HyyGvMz7HiihS86aBdzATi1V807CdrJc1TWZ+ztVXj7uk0XeI/zMFDordQs
Q3MMQ0Mf0CKGIVIGDK6APndR2+w=
=nZnF
-----END PGP MESSAGE-----
```
</details>

is also interested in backing up issues more cleanly that regularly one-time migrating them.

He made a script that he should send me by 19/02/25 evening automating migrating, he compares last 2 commits, issues and releases to avoid any overwrite of local. Well we discussed again about it on 26/02/25 there is a new version the person is working on having issues, issue comments and commits support (previously it treated a single branch) but it behaves well in a not malicious environment (for instance not all issues except most recent one being deleted).

<details>
<summary>He shared his script at:</summary>

```
-----BEGIN PGP MESSAGE-----

hF4DTQa9Wom5MBgSAQdAvQZk0ZIs2768aUpM23PkiV7+PD8dAGlCFACzsay4RiUw
83B7BHVsgwY2PozkhpI3qxKj0WsZA3KruCkuvShJlq6PU4v0fNIxujWIXjQVt96e
0nABuhX6qt6zA2Mj+oFiidA9NVWnU2K0YaVBMGsXfY+Dufo3YsLr7xbMjhsO81q3
JRs6EA8l2yRUtX4Vfe95J3Ow566xrspuKlJsjr9ufQMRJzSWVe6Y+UftzxtSrdMF
Uh7dV/FRljoNs5U1Z3r8E+tj
=u8JF
-----END PGP MESSAGE-----
```
</details>

https://gitea.lemnoslife.com/api/swagger#/issue/issueCreateIssue

does not give list being able to impersonate the author or the date time even if connected with an administrator account.
Maybe changing this part of the source code looks easier than making realtime migration but then still have import from JSON to do.

Exporting new issues changes since last backup seems the most appropriate if not realtime.

According to the mentioned person we can retrieve the issue history with the API but maybe he meant issues since a given date.

**Comment by Benjamin-Loison** [comment]
An alternative is to commit issues within the repository itself as Markdown files, like current [Benjamin_Loison/Food/issues](https://codeberg.org/Benjamin_Loison/Food/issues).

**Comment by Benjamin-Loison** [comment]
[Benjamin_Loison/nginx/settings/hooks](https://gitea.lemnoslife.com/Benjamin_Loison/nginx/settings/hooks)
https://docs.gitea.com/usage/webhooks

may help for realtime.

**Comment by Benjamin-Loison** [comment]
See [Benjamin_Loison/codeberg/issues/9](https://gitea.lemnoslife.com/Benjamin_Loison/codeberg/issues/9).

**Comment by Benjamin-Loison** [comment]
Related to [Benjamin_Loison/gitlab/issues/21](https://gitlab.com/Benjamin_Loison/gitlab/-/issues/21).

**Comment by Benjamin-Loison** [comment]
Related to [Improve_websites_thanks_to_open_source/issues/679](https://codeberg.org/Benjamin_Loison/Improve_websites_thanks_to_open_source/issues/679).

**Comment by Benjamin-Loison** [comment]
> From time to time, we check how people use our resources. When we think that projects use an unreasonable or otherwise needlessly excessive amount of resources, we handle such situations on a case-by-case basis, attempting to contact the projects' maintainers and/or enforcing limits.

Source: https://blog.codeberg.org/new-storage-limits-on-codeberg-what-you-need-to-know.html

I read it all.

**Comment by Benjamin-Loison** [comment]
My images being removed make me think that we can potentially provide links to other websites instead of Codeberg and these are loaded and we get IPs of people? Well maybe #76 protects against that.

**Comment by Benjamin-Loison** [comment]
<details>
<summary>The person:</summary>

```
-----BEGIN PGP MESSAGE-----

hF4DTQa9Wom5MBgSAQdAce1Z9M0sn1SzBh10HTA/+xvYx0Au9bX4Vd89Io1A7lQw
YsWPkOku8eNML3kEI7ZLfGOGlJfsKLjte4Q+lSlb+UaidSQCGMwAekAs6ulhJAIH
1EwBCQIQhT8VpIGF4vL072YDqsbwVzxJ+stckYZYYdG9u/Jo+wS2GM94qAae/rik
Ue7rA/+/xcdX3mFVEYzYAj2Y8oaabevBBfBNZvyr
=pWkh
-----END PGP MESSAGE-----
```
</details>

does not know a tool to migrate more than once issues, or just export Gitea issues as asked orally on 08/09/2026. Maybe this statement is considering mirroring issues from Gitea to GitHub.

Related to [Improve_websites_thanks_to_open_source/issues/2366](https://codeberg.org/Benjamin_Loison/Improve_websites_thanks_to_open_source/issues/2366).

<details>
<summary>See the message asking him again:</summary>

```
-----BEGIN PGP MESSAGE-----

hF4DTQa9Wom5MBgSAQdA70ox2PSZhAvpRiaTIOpf6WJtQMjME4bbfeBjKN+mNiMw
4Mt3JHNpemRiYXqfXRPLO+eNgas/OLp6CP4lxw6rq7CvxJu6zoI/DunDWl4JR9yk
1LQBCQIQwJXstaIPX1jrAAsecqQdcFXTcXRtejDQwe7NdwW3IXQ3I8Cqf0w8iC2e
BxI1NMAzoT8FMlyuw3m7+uOLSFeifN5VMQUymrgTZE4k46T8F1f/6GYGRoLQJ3jO
CTVpn4ghCI+1Brn1NpcgQZVVnTY3qXJt0nt0rNfFtU3b9RaL9PhsgqRIthc1aNjp
yIR2CXoF9HXVmPdUiECkFGUxSgDV/DEb7kTz2OXNUWSMG9WdunE=
=upfz
-----END PGP MESSAGE-----
```
</details>

He said that he does not know how to push issues from Gitea to GitHub.

**Comment by Benjamin-Loison** [comment]
Are they opposed to give limited rate read-only partial database access? Possibly to a database copy, specifying when the copy happens to avoid corruption or keep last both, assuming that single SQL query takes less than a copy cache period.

**Comment by Benjamin-Loison** [comment]
https://docs.codeberg.org/git/reduce-size/
https://docs.codeberg.org/getting-started/faq/#i-removed-an-object-from-my-repo%2C-why-doesn't-the-reported-size-shrink%3F

Source: https://matrix.to/#/!ImcWflUWyvAekDFLHt:matrix.tu-berlin.de/$0Abi7yOD-z9V6wn58uYDPwHwQeiv1b1YMouzaxR6R0I

**Comment by Benjamin-Loison** [comment]
I have read []go-gitea/gitea/issues/4351(https://github.com/go-gitea/gitea/issues/4351) which does not help.

**Comment by Benjamin-Loison** [comment]
Related to [Benjamin_Loison/codeberg/issues/19](https://gitea.lemnoslife.com/Benjamin_Loison/codeberg/issues/19).

**Comment by Benjamin-Loison** [comment]
DuckDuckGo and Google search *Codeberg backup*.

https://chat.mistral.ai/chat/494987c3-c530-46ca-a453-eb5c5039e8f1

https://docs.codeberg.org/getting-started/faq/#where-is-codeberg-hosted%3F

> Most of our services run on our own hardware in a rented facility in Berlin, Germany. The networking and rack space is provided by another volunteer-driven association.
>
> Certain tasks are offloaded to other locations and providers, e.g. [netcup GmbH](https://www.netcup.de/) and [Hetzner Online GmbH](https://www.hetzner.com/). This is for backups, redundancy, disaster recovery, DDoS protection, efficiency (using spare resources provided to us) and various other reasons.

**Comment by Benjamin-Loison** [comment]
https://forum.gitea.com/t/script-to-download-all-repos-for-one-user/6744 may help.

**Comment by Benjamin-Loison** [comment]
Related to [Benjamin_Loison/Heat/issues/13](https://codeberg.org/Benjamin_Loison/Heat/issues/13).

**Comment by Benjamin-Loison** [comment]
<details>
<summary>Output:</summary>

```
asset:  11%|█         | 1/9 [00:46<06:10, 46.33s/it]
Traceback (most recent call last):
  File "/home/benjamin_loison/venv/lib/python3.14/site-packages/urllib3/response.py", line 905, in _error_catcher
    yield
  File "/home/benjamin_loison/venv/lib/python3.14/site-packages/urllib3/response.py", line 1043, in _raw_read
    raise IncompleteRead(self._fp_bytes_read, self.length_remaining)
urllib3.exceptions.IncompleteRead: IncompleteRead(98304 bytes read, 76 more expected)

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/home/benjamin_loison/Desktop/bens_folder/dev/python/projects/backup_gitea_issue_attachments/main.py", line 34, in <module>
    downloadAssets(assets)
    ~~~~~~~~~~~~~~^^^^^^^^
  File "/home/benjamin_loison/Desktop/bens_folder/dev/python/projects/backup_gitea_issue_attachments/main.py", line 23, in downloadAssets
    shutil.copyfileobj(r.raw, f)
    ~~~~~~~~~~~~~~~~~~^^^^^^^^^^
  File "/home/benjamin_loison/.local/share/uv/python/cpython-3.14.6-linux-x86_64-gnu/lib/python3.14/shutil.py", line 256, in copyfileobj
    while buf := fsrc_read(length):
                 ~~~~~~~~~^^^^^^^^
  File "/home/benjamin_loison/venv/lib/python3.14/site-packages/urllib3/response.py", line 1110, in read
    data = self._raw_read(amt)
  File "/home/benjamin_loison/venv/lib/python3.14/site-packages/urllib3/response.py", line 1021, in _raw_read
    with self._error_catcher():
         ~~~~~~~~~~~~~~~~~~~^^
  File "/home/benjamin_loison/.local/share/uv/python/cpython-3.14.6-linux-x86_64-gnu/lib/python3.14/contextlib.py", line 162, in __exit__
    self.gen.throw(value)
    ~~~~~~~~~~~~~~^^^^^^^
  File "/home/benjamin_loison/venv/lib/python3.14/site-packages/urllib3/response.py", line 925, in _error_catcher
    raise ProtocolError(arg, e) from e
urllib3.exceptions.ProtocolError: ('Connection broken: IncompleteRead(98304 bytes read, 76 more expected)', IncompleteRead(98304 bytes read, 76 more expected))
```
</details>

after

```bash
ls -lh 3244af39-6d3d-4458-881d-28b63f2d6da4 
```
```
-rw-rw-r-- 1 benjamin_loison benjamin_loison 74K Sep  6 16:01 3244af39-6d3d-4458-881d-28b63f2d6da4
```

It should discard write so far in my opinion.

<details>
<summary>Python script:</summary>

```python
# [Benjamin-Loison/gitea/issues/46](https://github.com/Benjamin-Loison/gitea/issues/46)

import requests
from tqdm import tqdm
import os
import shutil
import time

REPOSITORY_FULL_NAME = 'Benjamin_Loison/Voice_assistant'
ISSUE_ID = None
COMMENT_ID = 22425157
ATTACHMENTS_FOLDER = '/home/benjamin/Desktop/bens_folder/dev/git/codeberg/attachments/'

def getApi(url):
    return requests.get(f'https://codeberg.org/api/v1/{url}').json()

def downloadAssets(assets):
    for asset in tqdm(assets, desc = 'asset'):
        uuid = asset['uuid']
        filePath = f'{ATTACHMENTS_FOLDER}/{uuid}'
        if not os.path.isfile(filePath):
            while True:
                toSleep = 1
                try:
                    with requests.get(asset['browser_download_url'], stream = True) as r:
                        with open(filePath, 'wb') as f:
                            shutil.copyfileobj(r.raw, f)
                except urllib3.exceptions.ProtocolError:
                    print(f'Sleeping {toSleep} seconds')
                    time.sleep(toSleep)
                    print(f'Finished sleeping')
                    toSleep *= 2
        else:
            print(f'{uuid} already exists!')

if ISSUE_ID is not None:
    assets = getApi(f'repos/{REPOSITORY_FULL_NAME}/issues/{ISSUE_ID}/assets')
if COMMENT_ID is not None:
    comment = getApi(f'repos/{REPOSITORY_FULL_NAME}/issues/comments/{COMMENT_ID}')
    assets = comment['assets']

downloadAssets(assets)
```
</details>

hangs without outputting anything, so I suspect facing time out.

**Comment by Benjamin-Loison** [comment]
<details>
<summary>If <i>Interrupt the running code</i>:</summary>

```
Traceback (most recent call last):
  File "/home/benjamin_loison/Desktop/bens_folder/dev/python/projects/backup_gitea_issue_attachments/main.py", line 25, in downloadAssets
    with requests.get(asset['browser_download_url'], stream = True) as r:
         ~~~~~~~~~~~~^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/benjamin_loison/venv/lib/python3.14/site-packages/requests/api.py", line 87, in get
    return request("get", url, params=params, **kwargs)
  File "/home/benjamin_loison/venv/lib/python3.14/site-packages/requests/api.py", line 71, in request
    return session.request(method=method, url=url, **kwargs)
           ~~~~~~~~~~~~~~~^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/benjamin_loison/venv/lib/python3.14/site-packages/requests/sessions.py", line 651, in request
    resp = self.send(prep, **send_kwargs)
  File "/home/benjamin_loison/venv/lib/python3.14/site-packages/requests/sessions.py", line 784, in send
    r = adapter.send(request, **kwargs)
  File "/home/benjamin_loison/venv/lib/python3.14/site-packages/requests/adapters.py", line 696, in send
    resp = conn.urlopen(
        method=request.method,
    ...<9 lines>...
        chunked=chunked,
    )
  File "/home/benjamin_loison/venv/lib/python3.14/site-packages/urllib3/connectionpool.py", line 788, in urlopen
    response = self._make_request(
        conn,
    ...<10 lines>...
        **response_kw,
    )
  File "/home/benjamin_loison/venv/lib/python3.14/site-packages/urllib3/connectionpool.py", line 464, in _make_request
    self._validate_conn(conn)
    ~~~~~~~~~~~~~~~~~~~^^^^^^
  File "/home/benjamin_loison/venv/lib/python3.14/site-packages/urllib3/connectionpool.py", line 1106, in _validate_conn
    conn.connect()
    ~~~~~~~~~~~~^^
  File "/home/benjamin_loison/venv/lib/python3.14/site-packages/urllib3/connection.py", line 759, in connect
    self.sock = sock = self._new_conn()
                       ~~~~~~~~~~~~~~^^
  File "/home/benjamin_loison/venv/lib/python3.14/site-packages/urllib3/connection.py", line 204, in _new_conn
    sock = connection.create_connection(
        (self._dns_host, self.port),
    ...<2 lines>...
        socket_options=self.socket_options,
    )
  File "/home/benjamin_loison/venv/lib/python3.14/site-packages/urllib3/util/connection.py", line 73, in create_connection
    sock.connect(sa)
    ~~~~~~~~~~~~^^^^
KeyboardInterrupt

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/home/benjamin_loison/Desktop/bens_folder/dev/python/projects/backup_gitea_issue_attachments/main.py", line 43, in <module>
    downloadAssets(assets)
    ~~~~~~~~~~~~~~^^^^^^^^
  File "/home/benjamin_loison/Desktop/bens_folder/dev/python/projects/backup_gitea_issue_attachments/main.py", line 28, in downloadAssets
    except urllib3.exceptions.ProtocolError:
           ^^^^^^^
NameError: name 'urllib3' is not defined
```
</details>

**Comment by Benjamin-Loison** [comment]
<details>
<summary>So now use Python script:</summary>

```python
# [Benjamin-Loison/gitea/issues/46](https://github.com/Benjamin-Loison/gitea/issues/46)

import requests
from tqdm import tqdm
import os
import shutil
import time
import urllib3

REPOSITORY_FULL_NAME = 'Benjamin_Loison/Voice_assistant'
ISSUE_ID = None
COMMENT_ID = 22425157
ATTACHMENTS_FOLDER = '/home/benjamin/Desktop/bens_folder/dev/git/codeberg/attachments/'

def getApi(url):
    return requests.get(f'https://codeberg.org/api/v1/{url}').json()

def downloadAssets(assets):
    for asset in tqdm(assets, desc = 'asset'):
        uuid = asset['uuid']
        filePath = f'{ATTACHMENTS_FOLDER}/{uuid}'
        if not os.path.isfile(filePath):
            while True:
                toSleep = 1
                try:
                    with requests.get(asset['browser_download_url'], stream = True) as r:
                        with open(filePath, 'wb') as f:
                            shutil.copyfileobj(r.raw, f)
                except urllib3.exceptions.ProtocolError:
                    print(f'Sleeping {toSleep} seconds')
                    time.sleep(toSleep)
                    print(f'Finished sleeping')
                    toSleep *= 2
        else:
            print(f'{uuid} already exists!')

if ISSUE_ID is not None:
    assets = getApi(f'repos/{REPOSITORY_FULL_NAME}/issues/{ISSUE_ID}/assets')
if COMMENT_ID is not None:
    comment = getApi(f'repos/{REPOSITORY_FULL_NAME}/issues/comments/{COMMENT_ID}')
    assets = comment['assets']

downloadAssets(assets)
```
</details>

**Comment by Benjamin-Loison** [comment]
<details>
<summary>If <i>Interrupt the running code</i>:</summary>

```
Traceback (most recent call last):
  File "/home/benjamin_loison/Desktop/bens_folder/dev/python/projects/backup_gitea_issue_attachments/main.py", line 44, in <module>
    downloadAssets(assets)
    ~~~~~~~~~~~~~~^^^^^^^^
  File "/home/benjamin_loison/Desktop/bens_folder/dev/python/projects/backup_gitea_issue_attachments/main.py", line 26, in downloadAssets
    with requests.get(asset['browser_download_url'], stream = True) as r:
         ~~~~~~~~~~~~^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/benjamin_loison/venv/lib/python3.14/site-packages/requests/api.py", line 87, in get
    return request("get", url, params=params, **kwargs)
  File "/home/benjamin_loison/venv/lib/python3.14/site-packages/requests/api.py", line 71, in request
    return session.request(method=method, url=url, **kwargs)
           ~~~~~~~~~~~~~~~^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/benjamin_loison/venv/lib/python3.14/site-packages/requests/sessions.py", line 651, in request
    resp = self.send(prep, **send_kwargs)
  File "/home/benjamin_loison/venv/lib/python3.14/site-packages/requests/sessions.py", line 784, in send
    r = adapter.send(request, **kwargs)
  File "/home/benjamin_loison/venv/lib/python3.14/site-packages/requests/adapters.py", line 696, in send
    resp = conn.urlopen(
        method=request.method,
    ...<9 lines>...
        chunked=chunked,
    )
  File "/home/benjamin_loison/venv/lib/python3.14/site-packages/urllib3/connectionpool.py", line 788, in urlopen
    response = self._make_request(
        conn,
    ...<10 lines>...
        **response_kw,
    )
  File "/home/benjamin_loison/venv/lib/python3.14/site-packages/urllib3/connectionpool.py", line 464, in _make_request
    self._validate_conn(conn)
    ~~~~~~~~~~~~~~~~~~~^^^^^^
  File "/home/benjamin_loison/venv/lib/python3.14/site-packages/urllib3/connectionpool.py", line 1106, in _validate_conn
    conn.connect()
    ~~~~~~~~~~~~^^
  File "/home/benjamin_loison/venv/lib/python3.14/site-packages/urllib3/connection.py", line 759, in connect
    self.sock = sock = self._new_conn()
                       ~~~~~~~~~~~~~~^^
  File "/home/benjamin_loison/venv/lib/python3.14/site-packages/urllib3/connection.py", line 204, in _new_conn
    sock = connection.create_connection(
        (self._dns_host, self.port),
    ...<2 lines>...
        socket_options=self.socket_options,
    )
  File "/home/benjamin_loison/venv/lib/python3.14/site-packages/urllib3/util/connection.py", line 73, in create_connection
    sock.connect(sa)
    ~~~~~~~~~~~~^^^^
KeyboardInterrupt
```
</details>

**Comment by Benjamin-Loison** [comment]
```python
len(requests.get('https://codeberg.org/Benjamin_Loison/Improve_websites_thanks_to_open_source/attachments/b3d0328e-d2c3-4246-ad59-f96bc9aa9d57').content)
```
```
138464
```

**Comment by Benjamin-Loison** [comment]
<details>
<summary>Well I have forgotten <code>break</code>, should use the Python script:</summary>

```python
# [Benjamin-Loison/gitea/issues/46](https://github.com/Benjamin-Loison/gitea/issues/46)

import requests
from tqdm import tqdm
import os
import shutil
import time
import urllib3

REPOSITORY_FULL_NAME = 'Benjamin_Loison/Improve_websites_thanks_to_open_source'
ISSUE_ID = 4055
COMMENT_ID = None
ATTACHMENTS_FOLDER = '/home/benjamin/Desktop/bens_folder/dev/git/codeberg/attachments/'

def getApi(url):
    return requests.get(f'https://codeberg.org/api/v1/{url}').json()

def downloadAssets(assets):
    for asset in tqdm(assets, desc = 'asset'):
        uuid = asset['uuid']
        filePath = f'{ATTACHMENTS_FOLDER}/{uuid}'
        if not os.path.isfile(filePath):
            while True:
                toSleep = 1
                try:
                    url = asset['browser_download_url']
                    print(f'{url=}')
                    with requests.get(url, stream = True) as r:
                        with open(filePath, 'wb') as f:
                            shutil.copyfileobj(r.raw, f)
                            break
                except urllib3.exceptions.ProtocolError:
                    print(f'Sleeping {toSleep} seconds')
                    time.sleep(toSleep)
                    print(f'Finished sleeping')
                    toSleep *= 2
        else:
            print(f'{uuid} already exists!')

if ISSUE_ID is not None:
    assets = getApi(f'repos/{REPOSITORY_FULL_NAME}/issues/{ISSUE_ID}/assets')
if COMMENT_ID is not None:
    comment = getApi(f'repos/{REPOSITORY_FULL_NAME}/issues/comments/{COMMENT_ID}')
    assets = comment['assets']

downloadAssets(assets)
```
</details>

**Comment by Benjamin-Loison** [comment]
I have removed the loop and will introduce it back when I face again the issue.
